api/metric/api.go

```diff
@@ -16,66 +16,97 @@ package metric
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// RecordAt allows the SDK to observe a single metric event
+	RecordAt(ctx context.Context, t time.Time, value float64)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, Instrument, ...core.KeyValue) Recorder
+
+	// RecordBatch atomically records a batch of measurements.
+	RecordBatch(context.Context, ...Measurement)
 }
 
-type Handle struct {
+// Instrument represents a named metric with recommended local-aggregation keys.
+type Instrument struct {
+	// Variable defines the metric name, description, and unit.
 	Variable registry.Variable
 
-	Type MetricType
+	// Kind is the metric kind of this instrument.
+	Kind Kind
+
+	// Keys are recommended keys determined in the handles
+	// obtained for this metric.
 	Keys []core.Key
 }
 
-type Option func(*Handle, *[]registry.Option)
+// Handle contains a Recorder to support the implementation-defined
+// behavior of reporting a single metric with pre-determined label
+// values.
+type Handle struct {
+	Recorder
+}
+
+// Measurement is used for reporting a batch of metric values.
+type Measurement struct {
+	Handle Handle
+	Value  float64
+}
+
+// Option supports specifying the various metric options.
+type Option func(*Instrument, *[]registry.Option)
 
 // WithDescription applies provided description.
 func WithDescription(desc string) Option {
-	return func(_ *Handle, to *[]registry.Option) {
+	return func(_ *Instrument, to *[]registry.Option) {
 		*to = append(*to, registry.WithDescription(desc))
 	}
 }
 
 // WithUnit applies provided unit.
 func WithUnit(unit unit.Unit) Option {
-	return func(_ *Handle, to *[]registry.Option) {
+	return func(_ *Instrument, to *[]registry.Option) {
 		*to = append(*to, registry.WithUnit(unit))
 	}
 }
 
-// WithKeys applies the provided dimension keys.
+// WithKeys applies recommended dimension keys.  Multiple `WithKeys`
+// options accumulate.  The keys specified in this way are taken as
+// the aggregation keys for Gauge and Cumulative.
 func WithKeys(keys ...core.Key) Option {
-	return func(m *Handle, _ *[]registry.Option) {
-		m.Keys = keys
+	return func(m *Instrument, _ *[]registry.Option) {
+		m.Keys = append(m.Keys, keys...)
 	}
 }
 
-func (mtype MetricType) String() string {
-	switch mtype {
-	case Gauge:
-		return "gauge"
-	case Cumulative:
-		return "cumulative"
-	default:
-		return "unknown"
-	}
+// Defined returns true when the instrument has been registered.
+func (inst Instrument) Defined() bool {
```

Comment:
Why does the Instrument need to exist?

Reply:
I'm not sure what you're asking. There are two kinds of thing here, one metric instrument (which does not depend on the SDK, is a plain struct) and one metric handle (which does depend on the SDK, has pre-defined label values, contains an interface to record with).